File: Algorithm/Go/700. Search in a Binary Search Tree.go
/*
 * Created on Thu Dec 20 2018 16:7:44
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func searchBST(root *TreeNode, val int) *TreeNode {
	if root == nil || root.Val == val {
		return root
	} else if root.Val > val {
		return searchBST(root.Left, val)
	} else {
		return searchBST(root.Right, val)
	}
}

File: Algorithm/C++/700. Search in a Binary Search Tree.cc
/*
 * Created on Thu Dec 20 2018 16:4:46
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
  TreeNode* searchBST(TreeNode* root, int val) {
    if (root == NULL || root->val == val)
      return root;
    else if (root->val > val)
      return searchBST(root->left, val);
    else
      return searchBST(root->right, val);
  }
};